File: sdks/swift/Sources/RustPipe/PipeWorker.swift
import Foundation

public typealias TaskHandler = @Sendable (PipeTask) async throws -> Any

public struct PipeTask: Sendable {
    public let id: String
    public let taskType: String
    public let payload: [String: Any]
}

public actor PipeWorker {
    private let url: URL
    private let workerId: String
    private let handlers: [String: TaskHandler]
    private let maxConcurrency: Int
    private let heartbeatInterval: TimeInterval
    private var webSocketTask: URLSessionWebSocketTask?
    private var activeTasks: Int = 0
    private let startTime = Date()

    public init(
        url: String,
        workerId: String? = nil,
        handlers: [String: TaskHandler],
        maxConcurrency: Int = 10,
        heartbeatInterval: TimeInterval = 5.0
    ) {
        self.url = URL(string: url)!
        self.workerId = workerId ?? "worker-swift-\(UUID().uuidString.prefix(8))"
        self.handlers = handlers
        self.maxConcurrency = maxConcurrency
        self.heartbeatInterval = heartbeatInterval
    }

    public func start() async throws {
        let session = URLSession(configuration: .default)
        webSocketTask = session.webSocketTask(with: url)
        webSocketTask?.resume()

        await register()
        startHeartbeat()
        await receiveLoop()
    }

    public func stop() {
        webSocketTask?.cancel(with: .normalClosure, reason: nil)
    }

    private func register() async {
        let msg: [String: Any] = [
            "type": "WorkerRegister",
            "registration": [
                "workerId": workerId,
                "supportedTasks": Array(handlers.keys),
                "maxConcurrency": maxConcurrency,
                "language": "Swift"
            ]
        ]
        await send(msg)
    }

    private func startHeartbeat() {
        Task {
            while webSocketTask?.state == .running {
                try await Task.sleep(nanoseconds: UInt64(heartbeatInterval * 1_000_000_000))
                let msg: [String: Any] = [
                    "type": "Heartbeat",
                    "payload": [
                        "workerId": workerId,
                        "activeTasks": activeTasks,
                        "capacity": maxConcurrency,
                        "uptimeSeconds": Int(Date().timeIntervalSince(startTime))
                    ]
                ]
                await send(msg)
            }
        }
    }

    private func receiveLoop() async {
        guard let ws = webSocketTask else { return }
        while ws.state == .running {
            do {
                let message = try await ws.receive()
                switch message {
                case .string(let text):
                    await handleMessage(text)
                default:
                    break
                }
            } catch {
                break
            }
        }
    }

    private func handleMessage(_ raw: String) async {
        guard let data = raw.data(using: .utf8),
              let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
              let type = json["type"] as? String else { return }

        switch type {
        case "TaskDispatch":
            if let taskDict = json["task"] as? [String: Any] {
                let task = PipeTask(
                    id: taskDict["id"] as? String ?? "",
                    taskType: taskDict["taskType"] as? String ?? "",
                    payload: taskDict["payload"] as? [String: Any] ?? [:]
                )
                Task { await executeTask(task) }
            }
        case "Shutdown":
            stop()
        default:
            break
        }
    }

    private func executeTask(_ task: PipeTask) async {
        activeTasks += 1
        let start = Date()

        defer { activeTasks -= 1 }

        guard let handler = handlers[task.taskType] else {
            await sendResult(taskId: task.id, status: "Failed", payload: nil,
                error: ["code": "UNKNOWN_TASK_TYPE", "message": "No handler for: \(task.taskType)", "retryable": false],
                durationMs: 0)
            return
        }

        do {
            let result = try await handler(task)
            let duration = Int(Date().timeIntervalSince(start) * 1000)
            await sendResult(taskId: task.id, status: "Completed", payload: result, error: nil, durationMs: duration)
        } catch {
            let duration = Int(Date().timeIntervalSince(start) * 1000)
            await sendResult(taskId: task.id, status: "Failed", payload: nil,
                error: ["code": "HANDLER_ERROR", "message": error.localizedDescription, "retryable": true],
                durationMs: duration)
        }
    }

    private func sendResult(taskId: String, status: String, payload: Any?, error: [String: Any]?, durationMs: Int) async {
        var result: [String: Any] = [
            "taskId": taskId,
            "status": status,
            "durationMs": durationMs,
            "workerId": workerId
        ]
        if let payload = payload { result["payload"] = payload }
        if let error = error { result["error"] = error }

        let msg: [String: Any] = ["type": "TaskResult", "result": result]
        await send(msg)
    }

    private func send(_ message: [String: Any]) async {
        guard let data = try? JSONSerialization.data(withJSONObject: message),
              let text = String(data: data, encoding: .utf8) else { return }
        try? await webSocketTask?.send(.string(text))
    }
}

File: sdks/java/src/main/java/io/rustpipe/PipeWorker.java
package io.rustpipe;

import com.google.gson.Gson;
import com.google.gson.JsonObject;
import org.java_websocket.client.WebSocketClient;
import org.java_websocket.handshake.ServerHandshake;

import java.net.URI;
import java.time.Instant;
import java.util.*;
import java.util.concurrent.*;
import java.util.concurrent.atomic.AtomicInteger;

public class PipeWorker {
    private final String url;
    private final String workerId;
    private final Map<String, TaskHandler> handlers;
    private final int maxConcurrency;
    private final long heartbeatIntervalMs;
    private final Gson gson = new Gson();
    private final ExecutorService executor;
    private final AtomicInteger activeTasks = new AtomicInteger(0);
    private final long startTime = System.currentTimeMillis();

    private WebSocketClient ws;
    private ScheduledExecutorService heartbeatScheduler;
    private volatile boolean running = false;

    public PipeWorker(Builder builder) {
        this.url = builder.url;
        this.workerId = builder.workerId != null ? builder.workerId :
                "worker-java-" + UUID.randomUUID().toString().substring(0, 8);
        this.handlers = builder.handlers;
        this.maxConcurrency = builder.maxConcurrency;
        this.heartbeatIntervalMs = builder.heartbeatIntervalMs;
        this.executor = Executors.newFixedThreadPool(maxConcurrency);
    }

    public void start() throws Exception {
        running = true;
        connect();
    }

    public void stop() {
        running = false;
        if (heartbeatScheduler != null) heartbeatScheduler.shutdown();
        if (ws != null) ws.close();
        executor.shutdown();
    }

    private void connect() throws Exception {
        ws = new WebSocketClient(new URI(url)) {
            @Override
            public void onOpen(ServerHandshake handshake) {
                register();
                startHeartbeat();
            }

            @Override
            public void onMessage(String raw) {
                handleMessage(raw);
            }

            @Override
            public void onClose(int code, String reason, boolean remote) {
                if (running) reconnect();
            }

            @Override
            public void onError(Exception ex) {
                System.err.println("[rust-pipe-java] Error: " + ex.getMessage());
            }
        };
        ws.connect();
    }

    private void reconnect() {
        try {
            Thread.sleep(1000);
            connect();
        } catch (Exception e) {
            System.err.println("[rust-pipe-java] Reconnect failed: " + e.getMessage());
        }
    }

    private void register() {
        JsonObject reg = new JsonObject();
        reg.addProperty("type", "WorkerRegister");

        JsonObject registration = new JsonObject();
        registration.addProperty("workerId", workerId);
        registration.add("supportedTasks", gson.toJsonTree(new ArrayList<>(handlers.keySet())));
        registration.addProperty("maxConcurrency", maxConcurrency);
        registration.addProperty("language", "Java");
        reg.add("registration", registration);

        ws.send(gson.toJson(reg));
    }

    private void startHeartbeat() {
        heartbeatScheduler = Executors.newSingleThreadScheduledExecutor();
        heartbeatScheduler.scheduleAtFixedRate(() -> {
            JsonObject hb = new JsonObject();
            hb.addProperty("type", "Heartbeat");

            JsonObject payload = new JsonObject();
            payload.addProperty("workerId", workerId);
            payload.addProperty("activeTasks", activeTasks.get());
            payload.addProperty("capacity", maxConcurrency);
            payload.addProperty("uptimeSeconds", (System.currentTimeMillis() - startTime) / 1000);
            hb.add("payload", payload);

            ws.send(gson.toJson(hb));
        }, heartbeatIntervalMs, heartbeatIntervalMs, TimeUnit.MILLISECONDS);
    }

    private void handleMessage(String raw) {
        JsonObject msg = gson.fromJson(raw, JsonObject.class);
        String type = msg.get("type").getAsString();

        switch (type) {
            case "TaskDispatch" -> {
                JsonObject taskJson = msg.getAsJsonObject("task");
                Task task = new Task(
                        taskJson.get("id").getAsString(),
                        taskJson.get("taskType").getAsString(),
                        taskJson.get("payload")
                );
                executor.submit(() -> executeTask(task));
            }
            case "Shutdown" -> stop();
        }
    }

    private void executeTask(Task task) {
        activeTasks.incrementAndGet();
        long start = System.currentTimeMillis();

        try {
            TaskHandler handler = handlers.get(task.taskType());
            if (handler == null) {
                sendResult(task.id(), "Failed", null,
                        new TaskError("UNKNOWN_TASK_TYPE", "No handler for: " + task.taskType(), false),
                        0);
                return;
            }

            Object result = handler.handle(task);
            long duration = System.currentTimeMillis() - start;
            sendResult(task.id(), "Completed", result, null, duration);
        } catch (Exception e) {
            long duration = System.currentTimeMillis() - start;
            sendResult(task.id(), "Failed", null,
                    new TaskError("HANDLER_ERROR", e.getMessage(), true), duration);
        } finally {
            activeTasks.decrementAndGet();
        }
    }

    private void sendResult(String taskId, String status, Object payload, TaskError error, long durationMs) {
        JsonObject msg = new JsonObject();
        msg.addProperty("type", "TaskResult");

        JsonObject result = new JsonObject();
        result.addProperty("taskId", taskId);
        result.addProperty("status", status);
        if (payload != null) result.add("payload", gson.toJsonTree(payload));
        if (error != null) result.add("error", gson.toJsonTree(error));
        result.addProperty("durationMs", durationMs);
        result.addProperty("workerId", workerId);
        msg.add("result", result);

        ws.send(gson.toJson(msg));
    }

    public static Builder builder(String url) {
        return new Builder(url);
    }

    public static class Builder {
        private final String url;
        private String workerId;
        private final Map<String, TaskHandler> handlers = new HashMap<>();
        private int maxConcurrency = 10;
        private long heartbeatIntervalMs = 5000;

        public Builder(String url) { this.url = url; }

        public Builder workerId(String id) { this.workerId = id; return this; }
        public Builder maxConcurrency(int n) { this.maxConcurrency = n; return this; }
        public Builder heartbeatInterval(long ms) { this.heartbeatIntervalMs = ms; return this; }
        public Builder handler(String taskType, TaskHandler h) { this.handlers.put(taskType, h); return this; }

        public PipeWorker build() { return new PipeWorker(this); }
    }

    public record Task(String id, String taskType, Object payload) {}
    public record TaskError(String code, String message, boolean retryable) {}

    @FunctionalInterface
    public interface TaskHandler {
        Object handle(Task task) throws Exception;
    }
}
